### PR TITLE
new: Add f1 metric and loss functions and GlobalF1 callback (fixes #172)

### LINKED ISSUE #172: Calculate loss based on f1
Something I think we have all been experiencing (especially when dealing with imbalanced classes) is that we evaluate our models based on f1 score, but we optimise using a loss like `binary_crossentropy`.

It would be great if we could evaluate and optimise for the same metric.

An example of how to create an f1 based loss is available here: https://www.kaggle.com/rejpalcz/best-loss-function-for-f1-score-metric#Optimal-loss-function---macro-F1-score

### PR DESCRIPTION
**Pull request by ivyleavedtoadflax**
## Description
Based on #172:

Adds two new functions:
* f1_metric: A metric for passing into keras `model.compile` which will return the batchwise macro f1 score
* f1_loss: A loss function which can be passed to `model.compile` which will optimise for the batchwise macro f1 score.

Expands the Metrics callback:
* Enables the global f1 history (against a validation set) to be saved for reporting purposes.

## Checklist

- [x] Added link to Github issue or Trello card
- [x] Added tests

### CODE REVIEW

File: wellcomeml/metrics/f1.py [hunk 1]
```diff
@@ -0,0 +1,113 @@
...
+    >>> global_f1 = GlobalF1(validation=(X_test, y_test, history.csv))
+    >>>
+    >>> history = model.fit(
+    >>>     X_train,
+    >>>     y_train,
+    >>>     epochs=100,
+    >>>     validation_data=(X_test, y_test),
+    >>>     batch_size=1024,
+    >>>     verbose=2,
+    >>>     callbacks=[early_stopping, glbal_f1],
```
Comment (aCampello) [inline wellcomeml/metrics/f1.py:71]: Typo

File: wellcomeml/metrics/f1.py [hunk 2]
```diff
@@ -0,0 +1,113 @@
...
+    p = tp / (tp + fp + K.epsilon())
+    r = tp / (tp + fn + K.epsilon())
+    f1 = 2 * p * r / (p + r + K.epsilon())
+    f1 = tf.where(tf.math.is_nan(f1), tf.zeros_like(f1), f1)
+
+    return 1 - K.mean(f1)
+
+
+class GlobalF1(Callback):
+    """Calculate global F1, precision, and recall, against a complete test set
```
Comment (aCampello) [inline wellcomeml/metrics/f1.py:59]: I think we might already have implemented this. The implementation is different but check out https://github.com/wellcometrust/WellcomeML/blob/master/wellcomeml/ml/keras_utils.py. Maybe your implementation is better, so you could change the one we already have instead of a new class?

**Review by aCampello — CHANGES_REQUESTED** [review]
Thanks a lot for the PR, Matt. I'm really keen to use the f1 loss and see whether this will improve performance to some of my existing projects!

One thing is that you ticked the "added tests" box, but I can't see them. Would be good to have tests to the new metric so we don't decrease coverage too much.

File: wellcomeml/metrics/f1.py (hunk 2 above)
Comment (ivyleavedtoadflax) [reply wellcomeml/metrics/f1.py:59]: ooh good spot!

**Comment by ivyleavedtoadflax** [comment]
> Thanks a lot for the PR, Matt. I'm really keen to use the f1 loss and see whether this will improve performance to some of my existing projects!
> 
> One thing is that you ticked the "added tests" box, but I can't see them. Would be good to have tests to the new metric so we don't decrease coverage too much.

oops yep will add them. I wondered why the test coverage was going down, not up!

File: wellcomeml/metrics/f1.py (hunk 2 above)
Comment (ivyleavedtoadflax) [reply wellcomeml/metrics/f1.py]: https://github.com/wellcometrust/WellcomeML/pull/171/commits/0637fe5bbfe094dbde91b56cc833d977a61fc08c

**Comment by ivyleavedtoadflax** [comment]
This should be ready to go now - it even bumped up overall test coverage by a tad!

File: tests/metrics/test_f1.py [hunk 3]
```diff
@@ -0,0 +1,120 @@
...
+
+
+@pytest.fixture(scope="module")
+def tmpdir(tmpdir_factory):
+    return tmpdir_factory.mktemp("test_f1")
+
+
+@pytest.fixture(scope="module")
+def data():
+    X_train = np.random.random((100, 10))
```
Comment (aCampello) [inline tests/metrics/test_f1.py:18]: This is a great test fixture. I would say it's not reproducible because of a random seed, but given the type of assertion below, I think it is fine.

File: wellcomeml/ml/keras_utils.py [hunk 4]
```diff
@@ -1,12 +1,34 @@
 from collections import defaultdict
 
-from sklearn.metrics import f1_score, precision_score, recall_score
+import pandas as pd
 import tensorflow as tf
+from sklearn.metrics import f1_score, precision_score, recall_score
 
 
 class Metrics(tf.keras.callbacks.Callback):
```
Comment (aCampello) [inline wellcomeml/ml/keras_utils.py:8]: This is nice. I am impressed that it worked with all the other classes that used it seamlessly

File: tests/metrics/test_f1.py (hunk 3 above)
Comment (ivyleavedtoadflax) [reply tests/metrics/test_f1.py:18]: Yes that's true, i think as you say, we are probably ok with the simple assertion I make for now. Perhaps there will need to be a time when we make a more concrete assertion and we could adjust the fixture.

File: wellcomeml/ml/keras_utils.py [hunk 5]
```diff
@@ -16,9 +38,26 @@ def on_epoch_end(self, epoch, logs):
...
+    def on_train_end(self, logs={}):
+        """Write metrics to csv file"""
+
+        if self.history_path:
+            history = {}
+            history["f1"] = self.f1s
+            history["precision"] = self.precisions
+            history["recall"] = self.recalls
+            history_df = pd.DataFrame(history)
+            history_df.to_csv(self.history_path, index_label="epoch")
```
Comment (aCampello) [inline wellcomeml/ml/keras_utils.py]: Pandas seems a bit of a cannon to kill a fly here. Can we save in jsonL?

**Review by aCampello — APPROVED** [review]
LGTM. Just wrote down some very minor comments, no need to address them if you don't want. :shipit:

**Comment by aCampello** [comment]
Your flake8 seems to have failed?

File: wellcomeml/ml/keras_utils.py (hunk 5 above)
Comment (ivyleavedtoadflax) [reply wellcomeml/ml/keras_utils.py]: Yeah this is true, and I had been thinking the same. Ideally we do want it in csv so that it can be easily read by dvc (the jsonl structure  required for this is a bit convoluted), let me switch this out and use the csv module instead.

File: wellcomeml/ml/keras_utils.py (hunk 4 above)
Comment (ivyleavedtoadflax) [reply wellcomeml/ml/keras_utils.py:8]: haha yes... :crossed_fingers:
